### PR TITLE
fix: argument of PodName function (fixes #7315)

### LINKED ISSUE #7315: Retrying failed workflow using WorkflowTemplate get stuck in Running phase of the workflow and Pending phase of the node  if POD_NAMES=v2
## Summary

What happened/what you expected to happen?

When retrying failed workflow, the old pod is not deleted and the new pod is not created if POD_NAMES=v2.
The UI tells the workflow's phases is Running and the node's phase is Pending. And nothing will change therefrom.

image

What version of Argo Workflows are you running?

master (d4aa9d1a6f308a59ec95bd0f0d6221fe899a6e06) (includes https://github.com/argoproj/argo-workflows/pull/7301)

## Diagnostics

examples for repro

```yaml
# templates.yaml
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: workflow-template-whalesay-template
spec:
  entrypoint: whalesay-template
  templates:
  - name: whalesay-template
    inputs:
      parameters:
      - name: message
    container:
      image: docker/whalesay
      command: [cowsay]
      args: ["{{inputs.parameters.message}}"]
---
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: workflow-template-fail-template
spec:
  templates:
  - name: fail-template
    container:
      image: python:alpine3.6
      command: [python, -c]
      # fail with a 100% probability
      args: ["import sys; sys.exit(1)"]
```

```yaml
# failure-in-step2.yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: workflow-template-retry-with-steps-
spec:
  entrypoint: retry-with-steps
  templates:
  - name: retry-with-steps
    steps:
    - - name: hello1
        templateRef:
          name: workflow-template-whalesay-template
          template: whalesay-template
        arguments:
          parameters:
          - name: message
            value: "I am hello1"
    - - name: hello2
        templateRef:
          name: workflow-template-fail-template
          template: fail-template
```

```
# fish
env POD_NAMES=v2 make start API=true UI=true
```

```
kubectl apply -f templates.yaml -n argo
argo submit failure-in-step2.yaml -n argo
```

What Kubernetes provider are you using?

k3d

What executor are you running?

Emissary? (just ran `make argoexec-image`)

```bash
# Logs from the server:
argo-server | time="2021-12-01T18:44:54.052Z" level=info msg="Deleting pod: workflow-template-retry-with-steps-qzzt8--2273160064"
```

Actual pod name is `workflow-template-retry-with-steps-qzzt8-fail-template-2273160064`

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by mikutas**
node has TemplateRef.Template field rather than TemplateName field

fixes #7315 

Signed-off-by: mikutas <23391543+mikutas@users.noreply.github.com>

Don't bother creating a PR until you've done this:

* [x] Run `make pre-commit -B` to fix codegen, lint, and commit message problems.

Create your PR as a draft.

* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it
  does not need to pass.
* Once required tests have passed, you can make it "Ready for review".
* Say how how you tested your changes. If you changed the UI, attach screenshots.

Tips:

* If changes were requested, and you've made them, then dismiss the review to get it looked at again.
* Add you organization to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md) if you like.
* You can ask for help!

### CODE REVIEW

File: workflow/util/util.go [hunk 1]
```diff
@@ -814,7 +814,13 @@ func retryWorkflow(ctx context.Context, kubeClient kubernetes.Interface, hydrato
 			return nil, errors.InternalErrorf("Workflow cannot be retried with node %s in %s phase", node.Name, node.Phase)
 		}
 		if node.Type == wfv1.NodeTypePod {
-			podName := PodName(wf.Name, node.Name, node.TemplateName, node.ID)
+			var templateName string
```
Comment (simster7) [inline workflow/util/util.go:817]: I see similar code here, but the template name format is different in the case of TemplateRef. Are we sure this is correct?

https://github.com/argoproj/argo-workflows/blob/7739256d4b28995e5af33d2e54c85322ed5dfcba/cmd/argo/commands/get.go#L532-L537

**Review by simster7 — DISMISSED** [review]
This needs tests

File: workflow/util/util.go (hunk 1 above)
Comment (mikutas) [reply workflow/util/util.go:817]: v2 pod name doesn't contain WorkflowTemplate's name (=`TemplateRef.name`).

image

It consists of [workflow name]-[template name]-[hash] .

---

Workflow
image

WorkflowTemplates
image

each full manifest is in #7315

File: workflow/util/util.go [hunk 2]
```diff
@@ -814,7 +814,13 @@ func retryWorkflow(ctx context.Context, kubeClient kubernetes.Interface, hydrato
...
+			if node.TemplateRef != nil {
+				templateName = node.TemplateRef.Template
+			} else {
+				templateName = node.TemplateName
+			}
+			podName := PodName(wf.Name, node.Name, templateName, node.ID)
```
Comment (alexec) [inline workflow/util/util.go:823]: This could do with a test.

**Comment by alexec** [comment]
@JPZ13 could I ask you to own the review?

**Comment by JPZ13** [comment]
@alexec Happy to. I'll give it a look this afternoon. Thanks for finding the bug and turning around a fix @mikutas

File: workflow/util/util.go (hunk 1 above)
Comment (JPZ13) [reply workflow/util/util.go:817]: Hey @mikutas - I'd recommend using the utility function for getting the template name: https://github.com/argoproj/argo-workflows/blob/master/ui/src/app/shared/pod-name.ts#L49

That utility function could use a test though. Feel free to add one in this PR

File: workflow/util/util.go [hunk 3]
```diff
@@ -866,6 +867,13 @@ func retryWorkflow(ctx context.Context, kubeClient kubernetes.Interface, hydrato
 	return wfClient.Update(ctx, newWF, metav1.UpdateOptions{})
 }
 
+func getTemplateFromNode(node wfv1.NodeStatus) string {
```
Comment (JPZ13) [inline workflow/util/util.go:870]: Looks good! Could you add a test? In the meantime, I'll run it locally to test it out

**Comment by JPZ13** [comment]
I've confirmed this fix works when running locally. I'll approve once there's a unit test for `getTemplateFromNode`

**Review by JPZ13 — APPROVED** [review]
Looks great!

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
This will need back porting to v3.2

**Pull request merged**